Ubuntu 18.10 Build Warnings

0. Reviewed guide and contributing documents? (Yes/No): Yes
1. version [x.y.z, hash, other]: 4.2.1
2. installed as a package or compiled from sources [deb, rpm, git, other]: git
3. standalone or part of third party [motion, MotionEyeOS, other]: standalone
4. video stream source [V4L (card or USB), net cam (mjpeg, rtsp, other), mmal]: rtsp
5. hardware [x86, ARM, other]: x86
6. operating system [16.04, Stretch, etc, FreeBSD, other]: 18.10

Resolve the following build warnings on 18.10 and add 18.10 to the travis tests:
```
/motion/motion# make
Welcome to the setup procedure for Motion, the motion detection daemon! If you get
error messages during this procedure, please report them to the mailing list. The
Motion Guide contains all information you should need to get Motion up and running.

Version: 4.2.1+git20181209-7bd179c
Platform: Linux (if this is incorrect, please read README.FreeBSD)

Generating dependencies, please wait...

Compiling Motion object files...
--------------------------------------------------------------------------------
-e 	Compiling motion.c into motion.o...
-e 	Compiling logger.c into logger.o...
-e 	Compiling conf.c into conf.o...
conf.c: In function 'conf_load':
conf.c:2362:41: warning: '/motion.conf' directive output may be truncated writing 12 bytes into a region of size between 1 and 4096 [-Wformat-truncation=]
         snprintf(filename, PATH_MAX, "%s/motion.conf", path);
                                         ^~~~~~~~~~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from conf.c:30:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output between 13 and 4108 bytes into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-e 	Compiling draw.c into draw.o...
-e 	Compiling jpegutils.c into jpegutils.o...
-e 	Compiling video_loopback.c into video_loopback.o...
-e 	Compiling video_v4l2.c into video_v4l2.o...
-e 	Compiling video_common.c into video_common.o...
video_common.c: In function 'vid_greytoyuv420p':
video_common.c:499:45: warning: integer overflow in expression '-4784 * (int)*r - 9437 * (int)*r + 14221 * (int)*r' of type 'int' results in '0' [-Woverflow]
             *u += ((-4784 ** r - 9437 **  r + 14221 ** r) >> 17) + 32;
                    ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
video_common.c:504:45: warning: integer overflow in expression '-4784 * (int)*r - 9437 * (int)*r + 14221 * (int)*r' of type 'int' results in '0' [-Woverflow]
             *u += ((-4784 ** r - 9437 **  r + 14221 ** r) >> 17) + 32;
                    ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
-e 	Compiling video_bktr.c into video_bktr.o...
-e 	Compiling netcam.c into netcam.o...
-e 	Compiling netcam_http.c into netcam_http.o...
-e 	Compiling netcam_ftp.c into netcam_ftp.o...
-e 	Compiling netcam_jpeg.c into netcam_jpeg.o...
-e 	Compiling netcam_wget.c into netcam_wget.o...
-e 	Compiling track.c into track.o...
-e 	Compiling alg.c into alg.o...
-e 	Compiling event.c into event.o...
event.c: In function 'event_ffmpeg_newfile':
event.c:946:60: warning: '%s' directive output may be truncated writing up to 4095 bytes into a region of size between 4083 and 4087 [-Wformat-truncation=]
         snprintf(cnt->motionfilename, PATH_MAX - 4, "%s/%s_%sm", cnt->conf.target_dir, codec, stamp);
                                                            ^~                                 ~~~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 7 or more bytes (assuming 4106) into a destination of size 4092
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c:947:57: warning: '%s' directive output may be truncated writing up to 4095 bytes into a region of size between 4083 and 4087 [-Wformat-truncation=]
         snprintf(cnt->newfilename, PATH_MAX - 4, "%s/%s_%s", cnt->conf.target_dir, codec, stamp);
                                                         ^~                                ~~~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 6 or more bytes (assuming 4105) into a destination of size 4092
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c:949:57: warning: '%s' directive output may be truncated writing up to 4095 bytes into a region of size 4091 [-Wformat-truncation=]
         snprintf(cnt->motionfilename, PATH_MAX - 4, "%s/%sm", cnt->conf.target_dir, stamp);
                                                         ^~                          ~~~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 3 or more bytes (assuming 4098) into a destination of size 4092
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c:950:54: warning: '%s' directive output may be truncated writing up to 4095 bytes into a region of size 4091 [-Wformat-truncation=]
         snprintf(cnt->newfilename, PATH_MAX - 4, "%s/%s", cnt->conf.target_dir, stamp);
                                                      ^~                         ~~~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 2 or more bytes (assuming 4097) into a destination of size 4092
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c: In function 'event_ffmpeg_timelapse':
event.c:1059:60: warning: '%s' directive output may be truncated writing up to 4095 bytes into a region of size 4091 [-Wformat-truncation=]
         snprintf(cnt->timelapsefilename, PATH_MAX - 4, "%s/%s", cnt->conf.target_dir, tmp);
                                                            ^~                         ~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 2 or more bytes (assuming 4097) into a destination of size 4092
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c: In function 'event_image_snapshot':
event.c:615:42: warning: '%s' directive output may be truncated writing between 3 and 4 bytes into a region of size between 0 and 4095 [-Wformat-truncation=]
         snprintf(filename, PATH_MAX, "%s.%s", filepath, imageext(cnt));
                                          ^~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output between 5 and 4101 bytes into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c:616:48: warning: '__builtin___snprintf_chk' output may be truncated before the last format character [-Wformat-truncation=]
         snprintf(fullfilename, PATH_MAX, "%s/%s", cnt->conf.target_dir, filename);
                                                ^
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 2 or more bytes (assuming 4097) into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c:596:42: warning: '%s' directive output may be truncated writing between 3 and 4 bytes into a region of size between 0 and 4095 [-Wformat-truncation=]
         snprintf(filename, PATH_MAX, "%s.%s", filepath, imageext(cnt));
                                          ^~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output between 5 and 4101 bytes into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c:597:48: warning: '__builtin___snprintf_chk' output may be truncated before the last format character [-Wformat-truncation=]
         snprintf(fullfilename, PATH_MAX, "%s/%s", cnt->conf.target_dir, filename);
                                                ^
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 2 or more bytes (assuming 4097) into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c: In function 'event_create_extpipe.part.7':
event.c:774:58: warning: '%s' directive output may be truncated writing up to 4095 bytes into a region of size 4091 [-Wformat-truncation=]
         snprintf(cnt->extpipefilename, PATH_MAX - 4, "%s/%s", cnt->conf.target_dir, stamp);
                                                          ^~                         ~~~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 2 or more bytes (assuming 4097) into a destination of size 4092
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c: In function 'event_image_preview.part.8':
event.c:690:51: warning: '.' directive output may be truncated writing 1 byte into a region of size between 0 and 4095 [-Wformat-truncation=]
             snprintf(previewname, PATH_MAX, "%s/%s.%s", cnt->conf.target_dir, filename, imageext(cnt));
                                                   ^
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 6 or more bytes (assuming 4102) into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c: In function 'event_imagem_detect.part.9':
event.c:561:43: warning: '__builtin___snprintf_chk' output may be truncated before the last format character [-Wformat-truncation=]
         snprintf(filenamem, PATH_MAX, "%sm", filename);
                                           ^
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output between 2 and 4097 bytes into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c:562:49: warning: '.' directive output may be truncated writing 1 byte into a region of size between 0 and 4095 [-Wformat-truncation=]
         snprintf(fullfilenamem, PATH_MAX, "%s/%s.%s", cnt->conf.target_dir, filenamem, imageext(cnt));
                                                 ^
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 6 or more bytes (assuming 4102) into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
event.c: In function 'event_image_detect.part.10':
event.c:524:48: warning: '.' directive output may be truncated writing 1 byte into a region of size between 0 and 4095 [-Wformat-truncation=]
         snprintf(fullfilename, PATH_MAX, "%s/%s.%s", cnt->conf.target_dir, filename, imageext(cnt));
                                                ^
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from event.c:10:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output 6 or more bytes (assuming 4102) into a destination of size 4096
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-e 	Compiling picture.c into picture.o...
picture.c: In function 'prepare_exif.constprop':
picture.c:249:38: warning: '%02d' directive output may be truncated writing between 2 and 11 bytes into a region of size between 9 and 16 [-Wformat-truncation=]
     snprintf(datetime_buf, 21, "%04d:%02d:%02d %02d:%02d:%02d",
                                      ^~~~
picture.c:249:32: note: directive argument in the range [-2147483647, 2147483647]
     snprintf(datetime_buf, 21, "%04d:%02d:%02d %02d:%02d:%02d",
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/stdio.h:873,
                 from /usr/include/postgresql/libpq-fe.h:23,
                 from motion.h:29,
                 from picture.h:13,
                 from picture.c:12:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: '__builtin___snprintf_chk' output between 20 and 72 bytes into a destination of size 21
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-e 	Compiling rotate.c into rotate.o...
-e 	Compiling translate.c into translate.o...
-e 	Compiling webu.c into webu.o...
-e 	Compiling webu_html.c into webu_html.o...
webu_html.c: In function 'webu_html_config':
webu_html.c:494:34: warning: passing argument 1 to restrict-qualified parameter aliases with argument 4 [-Wrestrict]
                         snprintf(response, sizeof (response),"%s%s", response, val_thread);
                                  ^~~~~~~~                            ~~~~~~~~
-e 	Compiling webu_text.c into webu_text.o...
-e 	Compiling webu_stream.c into webu_stream.o...
-e 	Compiling stream.c into stream.o...
-e 	Compiling md5.c into md5.o...
-e 	Compiling netcam_rtsp.c into netcam_rtsp.o...
-e 	Compiling ffmpeg.c into ffmpeg.o...
--------------------------------------------------------------------------------
Motion object files compiled.

Linking Motion...
--------------------------------------------------------------------------------
gcc  -o motion motion.o logger.o conf.o draw.o jpegutils.o video_loopback.o video_v4l2.o video_common.o video_bktr.o netcam.o netcam_http.o netcam_ftp.o netcam_jpeg.o netcam_wget.o track.o alg.o event.o picture.o rotate.o translate.o webu.o webu_html.o webu_text.o webu_stream.o stream.o md5.o netcam_rtsp.o ffmpeg.o  -lm  -pthread -ljpeg -lmicrohttpd  -lwebpmux -lwebp -lsqlite3 -lmysqlclient -lpq   -lavutil -lavformat -lavcodec -lswscale -lavdevice
--------------------------------------------------------------------------------
Motion has been linked.

Performing the conversion of .po to .mo files

Build complete, run "make install" to install Motion!
```

--- a/conf.c
+++ b/conf.c
@@ -3120,7 +3120,7 @@ static struct context **config_camera(struct context **cnt, const char *str,
  */
 static void usage()
 {
-    printf("motion Version "VERSION", Copyright 2000-2018 Jeroen Vreeken/Folkert van Heusden/Kenneth Lavrsen/Motion-Project maintainers\n");
+    printf("motion Version "VERSION", Copyright 2000-2019 Jeroen Vreeken/Folkert van Heusden/Kenneth Lavrsen/Motion-Project maintainers\n");
     printf("\nHome page :\t https://motion-project.github.io/ \n");
     printf("\nusage:\tmotion [options]\n");
     printf("\n\n");

--- a/conf.c
+++ b/conf.c
@@ -2359,7 +2359,9 @@ struct context **conf_load(struct context **cnt)
             exit(-1);
         }
 
-        snprintf(filename, PATH_MAX, "%s/motion.conf", path);
+        snprintf(filename, PATH_MAX, "%.*s/motion.conf"
+            , (int)(PATH_MAX-1-strlen("/motion.conf"))
+            , path);
         fp = fopen (filename, "r");
     }
 

--- a/event.c
+++ b/event.c
@@ -521,7 +521,11 @@ static void event_image_detect(struct context *cnt,
             imagepath = DEF_IMAGEPATH;
 
         mystrftime(cnt, filename, sizeof(filename), imagepath, currenttime_tv, NULL, 0);
-        snprintf(fullfilename, PATH_MAX, "%s/%s.%s", cnt->conf.target_dir, filename, imageext(cnt));
+        snprintf(fullfilename, PATH_MAX, "%.*s/%.*s.%s"
+            , (int)(PATH_MAX-2-strlen(filename)-strlen(imageext(cnt)))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-2-strlen(cnt->conf.target_dir)-strlen(imageext(cnt)))
+            , filename, imageext(cnt));
 
         passthrough = util_check_passthrough(cnt);
         if ((cnt->imgs.size_high > 0) && (!passthrough)) {
@@ -558,9 +562,14 @@ static void event_imagem_detect(struct context *cnt,
         mystrftime(cnt, filename, sizeof(filename), imagepath, currenttime_tv, NULL, 0);
 
         /* motion images gets same name as normal images plus an appended 'm' */
-        snprintf(filenamem, PATH_MAX, "%sm", filename);
-        snprintf(fullfilenamem, PATH_MAX, "%s/%s.%s", cnt->conf.target_dir, filenamem, imageext(cnt));
-
+        snprintf(filenamem, PATH_MAX, "%.*sm"
+            , (int)(PATH_MAX-1-strlen(filename))
+            , filename);
+        snprintf(fullfilenamem, PATH_MAX, "%.*s/%.*s.%s"
+            , (int)(PATH_MAX-2-strlen(filenamem)-strlen(imageext(cnt)))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-2-strlen(cnt->conf.target_dir)-strlen(imageext(cnt)))
+            , filenamem, imageext(cnt));
         put_picture(cnt, fullfilenamem, cnt->imgs.img_motion.image_norm, FTYPE_IMAGE_MOTION);
         event(cnt, EVENT_FILECREATE, NULL, fullfilenamem, (void *)FTYPE_IMAGE, currenttime_tv);
     }
@@ -593,8 +602,14 @@ static void event_image_snapshot(struct context *cnt,
             snappath = DEF_SNAPPATH;
 
         mystrftime(cnt, filepath, sizeof(filepath), snappath, currenttime_tv, NULL, 0);
-        snprintf(filename, PATH_MAX, "%s.%s", filepath, imageext(cnt));
-        snprintf(fullfilename, PATH_MAX, "%s/%s", cnt->conf.target_dir, filename);
+        snprintf(filename, PATH_MAX, "%.*s.%s"
+            , (int)(PATH_MAX-1-strlen(filepath)-strlen(imageext(cnt)))
+            , filepath, imageext(cnt));
+        snprintf(fullfilename, PATH_MAX, "%.*s/%.*s"
+            , (int)(PATH_MAX-1-strlen(filename))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-1-strlen(cnt->conf.target_dir))
+            , filename);
         put_picture(cnt, fullfilename, img_data->image_norm, FTYPE_IMAGE_SNAPSHOT);
         event(cnt, EVENT_FILECREATE, NULL, fullfilename, (void *)FTYPE_IMAGE, currenttime_tv);
 
@@ -602,7 +617,10 @@ static void event_image_snapshot(struct context *cnt,
          *  Update symbolic link *after* image has been written so that
          *  the link always points to a valid file.
          */
-        snprintf(linkpath, PATH_MAX, "%s/lastsnap.%s", cnt->conf.target_dir, imageext(cnt));
+        snprintf(linkpath, PATH_MAX, "%.*s/lastsnap.%s"
+            , (int)(PATH_MAX-strlen("/lastsnap.")-strlen(imageext(cnt)))
+            , cnt->conf.target_dir, imageext(cnt));
+
         remove(linkpath);
 
         if (symlink(filename, linkpath)) {
@@ -612,8 +630,14 @@ static void event_image_snapshot(struct context *cnt,
         }
     } else {
         mystrftime(cnt, filepath, sizeof(filepath), cnt->conf.snapshot_filename, currenttime_tv, NULL, 0);
-        snprintf(filename, PATH_MAX, "%s.%s", filepath, imageext(cnt));
-        snprintf(fullfilename, PATH_MAX, "%s/%s", cnt->conf.target_dir, filename);
+        snprintf(filename, PATH_MAX, "%.*s.%s"
+            , (int)(PATH_MAX-1-strlen(imageext(cnt)))
+            , filepath, imageext(cnt));
+        snprintf(fullfilename, PATH_MAX, "%.*s/%.*s"
+            , (int)(PATH_MAX-1-strlen(filename))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-1-strlen(cnt->conf.target_dir))
+            , filename);
         remove(fullfilename);
         put_picture(cnt, fullfilename, img_data->image_norm, FTYPE_IMAGE_SNAPSHOT);
         event(cnt, EVENT_FILECREATE, NULL, fullfilename, (void *)FTYPE_IMAGE, currenttime_tv);
@@ -687,7 +711,11 @@ static void event_image_preview(struct context *cnt,
                 imagepath = (char *)DEF_IMAGEPATH;
 
             mystrftime(cnt, filename, sizeof(filename), imagepath, &cnt->imgs.preview_image.timestamp_tv, NULL, 0);
-            snprintf(previewname, PATH_MAX, "%s/%s.%s", cnt->conf.target_dir, filename, imageext(cnt));
+            snprintf(previewname, PATH_MAX, "%.*s/%.*s.%s"
+                , (int)(PATH_MAX-2-strlen(filename)-strlen(imageext(cnt)))
+                , cnt->conf.target_dir
+                , (int)(PATH_MAX-2-strlen(cnt->conf.target_dir)-strlen(imageext(cnt)))
+                , filename, imageext(cnt));
 
             passthrough = util_check_passthrough(cnt);
             if ((cnt->imgs.size_high > 0) && (!passthrough)) {
@@ -771,7 +799,11 @@ static void event_create_extpipe(struct context *cnt,
         }
 
         mystrftime(cnt, stamp, sizeof(stamp), moviepath, currenttime_tv, NULL, 0);
-        snprintf(cnt->extpipefilename, PATH_MAX - 4, "%s/%s", cnt->conf.target_dir, stamp);
+        snprintf(cnt->extpipefilename, PATH_MAX - 4, "%.*s/%.*s"
+            , (int)(PATH_MAX-5-strlen(stamp))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-5-strlen(cnt->conf.target_dir))
+            , stamp);
 
         if (access(cnt->conf.target_dir, W_OK)!= 0) {
             /* Permission denied */
@@ -943,11 +975,27 @@ static void event_ffmpeg_newfile(struct context *cnt,
             codec = "msmpeg4";
             break;
         }
-        snprintf(cnt->motionfilename, PATH_MAX - 4, "%s/%s_%sm", cnt->conf.target_dir, codec, stamp);
-        snprintf(cnt->newfilename, PATH_MAX - 4, "%s/%s_%s", cnt->conf.target_dir, codec, stamp);
+        snprintf(cnt->motionfilename, PATH_MAX - 4, "%.*s/%s_%.*sm"
+            , (int)(PATH_MAX-7-strlen(stamp)-strlen(codec))
+            , cnt->conf.target_dir, codec
+            , (int)(PATH_MAX-7-strlen(cnt->conf.target_dir)-strlen(codec))
+            , stamp);
+        snprintf(cnt->newfilename, PATH_MAX - 4, "%.*s/%s_%.*s"
+            , (int)(PATH_MAX-6-strlen(stamp)-strlen(codec))
+            , cnt->conf.target_dir, codec
+            , (int)(PATH_MAX-6-strlen(cnt->conf.target_dir)-strlen(codec))
+            , stamp);
     } else {
-        snprintf(cnt->motionfilename, PATH_MAX - 4, "%s/%sm", cnt->conf.target_dir, stamp);
-        snprintf(cnt->newfilename, PATH_MAX - 4, "%s/%s", cnt->conf.target_dir, stamp);
+        snprintf(cnt->motionfilename, PATH_MAX - 4, "%.*s/%.*sm"
+            , (int)(PATH_MAX-6-strlen(stamp))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-6-strlen(cnt->conf.target_dir))
+            , stamp);
+        snprintf(cnt->newfilename, PATH_MAX - 4, "%.*s/%.*s"
+            , (int)(PATH_MAX-5-strlen(stamp))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-5-strlen(cnt->conf.target_dir))
+            , stamp);
     }
     if (cnt->conf.movie_output) {
         cnt->ffmpeg_output = mymalloc(sizeof(struct ffmpeg));
@@ -1056,7 +1104,11 @@ static void event_ffmpeg_timelapse(struct context *cnt,
         mystrftime(cnt, tmp, sizeof(tmp), timepath, currenttime_tv, NULL, 0);
 
         /* PATH_MAX - 4 to allow for .mpg to be appended without overflow */
-        snprintf(cnt->timelapsefilename, PATH_MAX - 4, "%s/%s", cnt->conf.target_dir, tmp);
+        snprintf(cnt->timelapsefilename, PATH_MAX - 4, "%.*s/%.*s"
+            , (int)(PATH_MAX-5-strlen(tmp))
+            , cnt->conf.target_dir
+            , (int)(PATH_MAX-5-strlen(cnt->conf.target_dir))
+            , tmp);
         passthrough = util_check_passthrough(cnt);
         cnt->ffmpeg_timelapse = mymalloc(sizeof(struct ffmpeg));
         if ((cnt->imgs.size_high > 0) && (!passthrough)){

--- a/motion.1
+++ b/motion.1
@@ -1,4 +1,4 @@
-.TH MOTION 1 2018-10-20 "Motion" "Motion Options and Config Files"
+.TH MOTION 1 2019-02-02 "Motion" "Motion Options and Config Files"
 .SH NAME
 motion \-   Detect motion using a video4linux device or network camera
 .SH SYNOPSIS
@@ -1800,6 +1800,19 @@ Description:
 .fi
 .RS
 Maximum frame rate to send to stream
+.RE
+.RE
+
+.TP
+.B stream_motion
+.RS
+.nf
+Values: on,off
+Default: off
+Description:
+.fi
+.RS
+Limit stream to 1 fps when no motion is being detected.
 .RE
 .RE
 

--- a/motion.c
+++ b/motion.c
@@ -364,7 +364,7 @@ static void sig_handler(int signo)
         if (cnt_list) {
             i = -1;
             while (cnt_list[++i]) {
-                cnt_list[i]->webcontrol_finish = 1;
+                cnt_list[i]->webcontrol_finish = TRUE;
                 cnt_list[i]->event_stop = TRUE;
                 cnt_list[i]->finish = 1;
                 /*
@@ -3435,6 +3435,8 @@ static void motion_watchdog(int indx){
      * Best to just not get into a watchdog situation...
      */
 
+    if (!cnt_list[indx]->running) return;
+
     cnt_list[indx]->watchdog--;
     if (cnt_list[indx]->watchdog == 0) {
         MOTION_LOG(ERR, TYPE_ALL, NO_ERRNO
@@ -3537,6 +3539,24 @@ static int motion_check_threadcount(void){
         if (cnt_list[indx]->running || cnt_list[indx]->restart)
             motion_threads_running++;
     }
+
+    /* If the web control/streams are in finish/shutdown, we
+     * do not want to count them.  They will be completely closed
+     * by the process outside of loop that is checking the counts
+     * of threads.  If the webcontrol is not in a finish / shutdown
+     * then we want to keep them in the tread count to allow user
+     * to restart the cameras and keep Motion running.
+     */
+    indx = 0;
+    while (cnt_list[indx] != NULL){
+        if ((cnt_list[indx]->webcontrol_finish == FALSE) &&
+            ((cnt_list[indx]->webcontrol_daemon != NULL) ||
+             (cnt_list[indx]->webstream_daemon != NULL))) {
+            motion_threads_running++;
+        }
+        indx++;
+    }
+
 
     if (((motion_threads_running == 0) && finish) ||
         ((motion_threads_running == 0) && (threads_running == 0))) {

--- a/motion_config.html
+++ b/motion_config.html
@@ -1174,6 +1174,12 @@
           <td align="left"><a href="#stream_maxrate" >stream_maxrate</a></td>
         </tr>
         <tr>
+          <td align="left">webcam_motion</td>
+          <td align="left">stream_motion</td>
+          <td align="left">stream_motion</td>
+          <td align="left"><a href="#stream_motion" >stream_motion</a></td>
+        </tr>
+        <tr>
           <td align="left">webcam_port</td>
           <td align="left">stream_port</td>
           <td align="left">stream_port</td>
@@ -1525,12 +1531,6 @@
           <td align="left">webcam_limit</td>
           <td align="left">stream_limit</td>
           <td align="left">stream_limit</td>
-          <td align="left">-Depreciated</td>
-        </tr>
-        <tr>
-          <td align="left">webcam_motion</td>
-          <td align="left">stream_motion</td>
-          <td align="left">stream_motion</td>
           <td align="left">-Depreciated</td>
         </tr>
         <tr>
@@ -1958,7 +1958,10 @@
               <td bgcolor="#edf4f9" ><a href="#stream_grey" >stream_grey</a> </td>
               <td bgcolor="#edf4f9" ><a href="#stream_maxrate" >stream_maxrate</a> </td>
             </tr>
-            </tbody>
+           <tr>
+              <td bgcolor="#edf4f9" ><a href="#stream_motion" >stream_motion</a> </td>
+           </tr>
+           </tbody>
         </table>
 
         <p></p>
@@ -4750,13 +4753,33 @@
           <li><code>{IP}:{port}/{camid}/action/eventend</code> Trigger the end of a event.</li>
           <li><code>{IP}:{port}/{camid}/action/snapshot</code> Create a snapshot</li>
           <li><code>{IP}:{port}/{camid}/action/restart</code> Shutdown and restart Motion</li>
-          <li><code>{IP}:{port}/{camid}/action/quit</code> Shutdown Motion</li>
+          <li><code>{IP}:{port}/{camid}/action/quit</code> Close all connections to the camera</li>
+          <li><code>{IP}:{port}/{camid}/action/end</code> Entirely shutdown the Motion application</li>
           <li><code>{IP}:{port}/{camid}/track/center</code> Send command to center PTZ camera</li>
           <li><code>{IP}:{port}/{camid}/track/set?x={value1}&y={value2}</code> Send command to PTZ camera to move
             to location specified by x and y</li>
           <li><code>{IP}:{port}/{camid}/track/set?pan={value1}&tilt={value2}</code> Send command to PTZ camera to
             pan to value1 and tilt to value2</li>
         </ul>
+        <p></p>
+        As a general rule, when the {camid} references the camera_id in the main motion.conf file, the webcontrol
+        actions referenced above are going to be applied to every camera that is connected to Motion.  This camera_id
+        is usually specified as 0 (zero).  So issuing a command of <code>{IP}:{port}/0/detection/pause</code> is
+        going to pause all the cameras.
+        <p></p>
+        A point of clarification with respect to the differences between <code>pause</code>, <code>quit</code>,
+        and <code>end</code>.  When the action of <code>pause</code> is executed, Motion will stop the motion detection
+        processing and of course all events but will continue to process and decode images from the camera.  This allows
+        for a faster transition when the user executes a <code>start</code>  The <code>quit</code> action conversely not
+        only stops the motion detection but also disconnects from the camera and decoding of images.  To start motion
+        detection after a <code>quit</code>, the user must execute a <code>restart</code> which will reinitialize the
+        connection to the camera.  And since the camera was completely disconnect, it can take more than a few seconds
+        for Motion to fully start and have the camera available for processing or viewing.  Finally, there is an
+        option for <code>end</code>.  This option completely terminates the Motion application.  It closes all connections
+        to all the cameras and terminates the application.  This may be required when running Motion in daemon mode.
+        Note that there is no way to restart the Motion application from the webcontrol interface after processing
+        a <code>end</code> request.
+
         <p></p>
         If the item above is available via the HTML/CSS interface, it is also possible to see the exact URL sent
         to Motion in the log.  Change the log level to 8 (debug), then open up the Motion webcontrol interface and
@@ -5196,6 +5219,18 @@
         Don't set this parameter too high unless you only use it on the localhost or on an internal LAN.
         <p></p>
 
+        <h3><a name="stream_motion"></a> stream_motion </h3>
+        <p></p>
+        <ul>
+          <li> Type: boolean</li>
+          <li> Range / Valid values: on, off</li>
+          <li> Default: off</li>
+        </ul>
+        <p></p>
+        Limit the framerate to 1 frame per second when there is no motion being detected and increase
+        it to the stream_maxrate when there is motion.
+        <p></p>
+
       </ul>
 
 
@@ -5388,8 +5423,8 @@
           <li> Default: Not defined</li>
         </ul>
         <p></p>
-        SQL query that executes on a event where a file is closed after a event of motion.
-        This can be used to for example update the events table with an end timestamp for the recordings.
+        SQL query that executes after a movie has finished.
+        This can be used to for example update the events table with an end timestamp for the recording.
         <p></p>
         <p></p>
         You can use <a href="#conversion_specifiers">Conversion Specifiers</a> within the query.

--- a/picture.c
+++ b/picture.c
@@ -235,6 +235,7 @@ static unsigned prepare_exif(unsigned char **exif,
     */
     char *description, *datetime, *subtime;
     char datetime_buf[22];
+    char tmpbuf[45];
     struct tm timestamp_tm;
     struct timeval tv1;
 
@@ -246,13 +247,19 @@ static unsigned prepare_exif(unsigned char **exif,
 
     localtime_r(&tv1.tv_sec, &timestamp_tm);
     /* Exif requires this exact format */
-    snprintf(datetime_buf, 21, "%04d:%02d:%02d %02d:%02d:%02d",
+    /* The compiler is twitchy on truncating formats and the exif is twitchy
+     * on the length of the whole string.  So we do it in two steps of printing
+     * into a large buffer which compiler wants, then print that into the smaller
+     * buffer that exif wants..TODO  Find better method
+     */
+    snprintf(tmpbuf, 45, "%04d:%02d:%02d %02d:%02d:%02d",
             timestamp_tm.tm_year + 1900,
             timestamp_tm.tm_mon + 1,
             timestamp_tm.tm_mday,
             timestamp_tm.tm_hour,
             timestamp_tm.tm_min,
             timestamp_tm.tm_sec);
+    snprintf(datetime_buf, 22,"%.21s",tmpbuf);
     datetime = datetime_buf;
 
     // TODO: Extract subsecond timestamp from somewhere, but only

--- a/video_common.c
+++ b/video_common.c
@@ -477,43 +477,9 @@ void vid_y10torgb24(unsigned char *map, unsigned char *cap_map, int width, int h
 
 void vid_greytoyuv420p(unsigned char *map, unsigned char *cap_map, int width, int height)
 {
-    /* This is a adaptation of the rgb to yuv.
-     * For grey, we use just a single color
-    */
 
-    unsigned char *y, *u, *v;
-    unsigned char *r;
-    int i, loop;
-
-    r = cap_map;
-
-    y = map;
-    u = y + width * height;
-    v = u + (width * height) / 4;
-    memset(u, 0, width * height / 4);
-    memset(v, 0, width * height / 4);
-
-    for (loop = 0; loop < height; loop++) {
-        for (i = 0; i < width; i += 2) {
-            *y++ = (9796 **  r + 19235 ** r + 3736 **  r) >> 15;
-            *u += ((-4784 ** r - 9437 **  r + 14221 ** r) >> 17) + 32;
-            *v += ((20218 ** r - 16941 ** r - 3277 **  r) >> 17) + 32;
-            r++;
-
-            *y++ = (9796 **  r + 19235 ** r + 3736 **  r) >> 15;
-            *u += ((-4784 ** r - 9437 **  r + 14221 ** r) >> 17) + 32;
-            *v += ((20218 ** r - 16941 ** r - 3277 **  r) >> 17) + 32;
-            r ++;
-
-            u++;
-            v++;
-        }
-
-        if ((loop & 1) == 0) {
-            u -= width / 2;
-            v -= width / 2;
-        }
-    }
+    memcpy(map, cap_map, (width*height));
+    memset(map+(width*height), 128, (width * height) / 2);
 
 }
 

--- a/webu.c
+++ b/webu.c
@@ -576,20 +576,52 @@ void webu_process_action(struct webui_ctx *webui) {
 
 
     } else if (!strcmp(webui->uri_cmd2,"restart")){
-        /* This is the legacy method...(we can do better than signals..).*/
         if (webui->thread_nbr == 0) {
-            MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO, _("httpd is going to restart"));
-            kill(getpid(),SIGHUP);
+            MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO, _("Restarting all threads"));
             webui->cntlst[0]->webcontrol_finish = TRUE;
+            kill(getpid(),SIGHUP);
         } else {
             MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO,
-                _("httpd is going to restart thread %d"),webui->thread_nbr);
+                _("Restarting thread %d"),webui->thread_nbr);
+            webui->cnt->restart = TRUE;
             if (webui->cnt->running) {
                 webui->cnt->event_stop = TRUE;
-                webui->cnt->finish = 1;
+                webui->cnt->finish = TRUE;
             }
-            webui->cnt->restart = 1;
+
         }
+
+    } else if (!strcmp(webui->uri_cmd2,"quit")){
+        if (webui->thread_nbr == 0 && webui->cam_threads > 1) {
+            while (webui->cntlst[++indx]){
+                MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO,
+                    _("Quitting thread %d"),webui->thread_nbr);
+                webui->cntlst[indx]->restart = FALSE;
+                webui->cntlst[indx]->event_stop = TRUE;
+                webui->cntlst[indx]->event_user = TRUE;
+                webui->cntlst[indx]->finish = TRUE;
+            }
+        } else {
+            MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO,
+                _("Quitting thread %d"),webui->thread_nbr);
+            webui->cnt->restart = FALSE;
+            webui->cnt->event_stop = TRUE;
+            webui->cnt->event_user = TRUE;
+            webui->cnt->finish = TRUE;
+        }
+
+    } else if (!strcmp(webui->uri_cmd2,"end")){
+            MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO, _("Motion terminating"));
+            while (webui->cntlst[indx]){
+                webui->cntlst[indx]->webcontrol_finish = TRUE;
+                webui->cntlst[indx]->restart = FALSE;
+                webui->cntlst[indx]->event_stop = TRUE;
+                webui->cntlst[indx]->event_user = TRUE;
+                webui->cntlst[indx]->finish = TRUE;
+                indx++;
+            }
+
+
     } else if (!strcmp(webui->uri_cmd2,"start")){
         if (webui->thread_nbr == 0 && webui->cam_threads > 1) {
             do {
@@ -2017,16 +2049,19 @@ void webu_stop(struct context **cnt) {
     int indxthrd;
 
     if (cnt[0]->webcontrol_daemon != NULL){
-        cnt[0]->webcontrol_finish = 1;
+        cnt[0]->webcontrol_finish = TRUE;
         MHD_stop_daemon (cnt[0]->webcontrol_daemon);
     }
+
 
     indxthrd = 0;
     while (cnt[indxthrd] != NULL){
         if (cnt[indxthrd]->webstream_daemon != NULL){
-            cnt[indxthrd]->webcontrol_finish = 1;
+            cnt[indxthrd]->webcontrol_finish = TRUE;
             MHD_stop_daemon (cnt[indxthrd]->webstream_daemon);
         }
+        cnt[indxthrd]->webstream_daemon = NULL;
+        cnt[indxthrd]->webcontrol_daemon = NULL;
         indxthrd++;
     }
 }
@@ -2037,6 +2072,7 @@ void webu_start(struct context **cnt) {
      * will not function correctly.
      */
     struct sigaction act;
+    int indxthrd;
 
     /* set signal handlers TO IGNORE */
     memset(&act, 0, sizeof(act));
@@ -2044,6 +2080,15 @@ void webu_start(struct context **cnt) {
     act.sa_handler = SIG_IGN;
     sigaction(SIGPIPE, &act, NULL);
     sigaction(SIGCHLD, &act, NULL);
+
+
+    indxthrd = 0;
+    while (cnt[indxthrd] != NULL){
+        cnt[indxthrd]->webstream_daemon = NULL;
+        cnt[indxthrd]->webcontrol_daemon = NULL;
+        cnt[indxthrd]->webcontrol_finish = FALSE;
+        indxthrd++;
+    }
 
     if (cnt[0]->conf.stream_preview_method != 99){
         webu_start_ports(cnt);

--- a/webu_html.c
+++ b/webu_html.c
@@ -493,7 +493,7 @@ static void webu_html_config(struct webui_ctx *webui) {
                         ,webui->cntlst[indx]->camera_id);
                     webu_write(webui, response);
                     if (val_thread != NULL){
-                        snprintf(response, sizeof (response),"%s%s", response, val_thread);
+                        snprintf(response, sizeof (response),"%s", val_thread);
                         webu_write(webui, response);
                     }
                 }

--- a/webu_html.c
+++ b/webu_html.c
@@ -324,6 +324,7 @@ static void webu_html_navbar_action(struct webui_ctx *webui) {
         "        <a onclick=\"action_click('/detection/pause');\">%s</a>\n"
         "        <a onclick=\"action_click('/detection/start');\">%s</a>\n"
         "        <a onclick=\"action_click('/action/restart');\">%s</a>\n"
+        "        <a onclick=\"action_click('/action/quit');\">%s</a>\n"
         "      </div>\n"
         "    </div>\n"
         ,_("Action")
@@ -335,7 +336,8 @@ static void webu_html_navbar_action(struct webui_ctx *webui) {
         ,_("Tracking")
         ,_("Pause")
         ,_("Start")
-        ,_("Restart"));
+        ,_("Restart")
+        ,_("Quit"));
     webu_write(webui, response);
 }
 

--- a/webu_text.c
+++ b/webu_text.c
@@ -435,22 +435,13 @@ static void webu_text_get_menu(struct webui_ctx *webui) {
 
 }
 
-static void webu_text_quit(struct webui_ctx *webui) {
+static void webu_text_action_quit(struct webui_ctx *webui) {
     /* Shut down motion or the associated thread */
     char response[WEBUI_LEN_RESP];
 
-    /* This is the legacy method...(we can do better than signals..).*/
-    if (webui->thread_nbr == 0) {
-        MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO, _("httpd quits"));
-        kill(getpid(),SIGQUIT);
-        webui->cntlst[0]->webcontrol_finish = TRUE;
-    } else {
-        MOTION_LOG(NTC, TYPE_STREAM, NO_ERRNO,
-            _("httpd quits thread %d"),webui->thread_nbr);
-        webui->cnt->restart = 0;
-        webui->cnt->event_stop = TRUE;
-        webui->cnt->finish = 1;
-    }
+    webu_process_action(webui);
+
+    webu_text_back(webui,"/action");
 
     webu_text_header(webui);
     snprintf(response,sizeof(response),
@@ -661,6 +652,10 @@ static void webu_text_action(struct webui_ctx *webui) {
     } else if (!strcmp(webui->uri_cmd2,"pause")){
         webu_text_action_pause(webui);
 
+    } else if ((!strcmp(webui->uri_cmd2,"quit")) ||
+               (!strcmp(webui->uri_cmd2,"end"))){
+        webu_text_action_quit(webui);
+
     } else if ((!strcmp(webui->uri_cmd2,"write")) ||
                (!strcmp(webui->uri_cmd2,"writeyes"))){
         webu_text_action_write(webui);
@@ -789,8 +784,9 @@ static void webu_text_menu_action(struct webui_ctx *webui) {
         "<a href=/%s/action/snapshot>snapshot</a><br>"
         "<a href=/%s/action/restart>restart</a><br>"
         "<a href=/%s/action/quit>quit</a><br>"
+        "<a href=/%s/action/end>end</a><br>"
         ,webui->uri_camid, webui->uri_camid, webui->uri_camid
-        ,webui->uri_camid, webui->uri_camid
+        ,webui->uri_camid, webui->uri_camid, webui->uri_camid
     );
     webu_write(webui, response);
     webu_text_trailer(webui);
@@ -1102,7 +1098,11 @@ void webu_text_main(struct webui_ctx *webui) {
 
     } else if ((strcmp(webui->uri_cmd1,"action") == 0) &&
                (strcmp(webui->uri_cmd2,"quit") == 0)){
-        webu_text_quit(webui);
+        webu_text_action(webui);
+
+    } else if ((strcmp(webui->uri_cmd1,"action") == 0) &&
+               (strcmp(webui->uri_cmd2,"end") == 0)){
+        webu_text_action(webui);
 
     } else if (!strcmp(webui->uri_cmd1,"action")) {
         webu_text_action(webui);

